crates/bevy_animation/src/curves.rs

```diff
@@ -0,0 +1,702 @@
+//! Curve structures used by the animation system.
+
+use bevy_math::{
+    curve::{cores::*, iterable::IterableCurve, *},
+    Quat, Vec3, Vec4, VectorSpace,
+};
+use bevy_reflect::Reflect;
+use thiserror::Error;
+
+/// A keyframe-defined curve that "interpolates" by stepping at `t = 1.0` to the next keyframe.
+#[derive(Debug, Clone, Reflect)]
+pub struct SteppedKeyframeCurve<T> {
+    core: UnevenCore<T>,
+}
+
+impl<T> Curve<T> for SteppedKeyframeCurve<T>
+where
+    T: Clone,
+{
+    #[inline]
+    fn domain(&self) -> Interval {
+        self.core.domain()
+    }
+
+    #[inline]
+    fn sample_unchecked(&self, t: f32) -> T {
+        self.core
+            .sample_with(t, |x, y, t| if t >= 1.0 { y.clone() } else { x.clone() })
+    }
+}
+
+impl<T> SteppedKeyframeCurve<T> {
+    /// Create a new [`SteppedKeyframeCurve`]. If the curve could not be constructed from the
+    /// given data, an error is returned.
+    #[inline]
+    pub fn new(timed_samples: impl IntoIterator<Item = (f32, T)>) -> Result<Self, UnevenCoreError> {
+        Ok(Self {
+            core: UnevenCore::new(timed_samples)?,
+        })
+    }
+}
+
+/// A keyframe-defined curve that uses cubic spline interpolation, backed by a contiguous buffer.
+#[derive(Debug, Clone, Reflect)]
+pub struct CubicKeyframeCurve<T> {
+    // Note: the sample width here should be 3.
+    core: ChunkedUnevenCore<T>,
+}
+
+impl<V> Curve<V> for CubicKeyframeCurve<V>
+where
+    V: VectorSpace,
+{
+    #[inline]
+    fn domain(&self) -> Interval {
+        self.core.domain()
+    }
+
+    #[inline]
+    fn sample_unchecked(&self, t: f32) -> V {
+        match self.core.sample_interp_timed(t) {
+            // In all the cases where only one frame matters, defer to the position within it.
+            InterpolationDatum::Exact((_, v))
+            | InterpolationDatum::LeftTail((_, v))
+            | InterpolationDatum::RightTail((_, v)) => v[1],
+
+            InterpolationDatum::Between((t0, u), (t1, v), s) => {
+                cubic_spline_interpolation(u[1], u[2], v[0], v[1], s, t1 - t0)
```

Comment:
I feel like this could use

- a) well named variables for `u[1]`, `u[2]`, `v[0]` and `v[1]` to add some context again
- b) a normal comment explaining what the values are

for future developers that look at the code in 1+ year even though (!!) it's already perfectly explained a few lines below in the `new` functions. imo we can't guarantee that people will find the other docs and if they don't this is harder to figure out.